### ISSUE TITLE
Add macro configurator

### ISSUE THREAD

**Issue by yurikhan** [yurikhan/teck#6]
The firmware contains code that looks like a macro engine, which seems to be able to play back any sequence of HID reports with specified delays in between.

Ability to configure the macros and bindings would nicely solve any feature requests that are not implementable in a way that works everywhere and for everyone.

**Comment by ChozoMufin** [comment]
I'd also really love to see this. Do you think it's possible?

**Comment by yurikhan** [comment]
If it were impossible, I would have happily closed this ticket long ago. The issue is that it is technically possible but very limiting — e.g. the firmware cannot know the currently active application, the operating system, and the logical keyboard layout(s) of the host.

**Comment by lydell** [comment]
> the firmware cannot know the currently active application, the operating system, and the logical keyboard layout(s) of the host.

I’d be perfectly happy with “dumb” macros which don’t know anything about those things. I’m mostly interested in creating macros that send, say first `shift` and then `1`. That might turn into a `!` on some systems, and something completely different on others, which is fine to me.

**Comment by ChozoMufin** [comment]
> I’d be perfectly happy with “dumb” macros which don’t know anything about those things. I’m mostly interested in creating macros that send, say first shift and then 1. That might turn into a ! on some systems, and something completely different on others, which is fine to me.

Same here!

**Comment by kkass** [comment]
I would agree with this request.  I use a programmer Dvorak layout, and I have to rely on software to remap some of the keys.  (This makes login after reboot very annoying if I have any numbers or special characters in my password.)

If you are not familiar, the programmer Dvorak layout changes the number row to make symbols the primary function, and numbers the shift function.  This would require exactly the behavior that lydell mentioned above.)
